src/details/ArborX_Ray.hpp

```diff
@@ -198,6 +198,16 @@ bool intersects(Ray const &ray, Box const &box)
   return intersection(ray, box, tmin, tmax) && (tmax >= 0.f);
 }
 
+KOKKOS_INLINE_FUNCTION
+float distance(Ray const &ray, Box const &box)
+{
+  float tmin;
+  float tmax;
+  bool intersects = intersection(ray, box, tmin, tmax) && (tmax >= 0.f);
+  return intersects ? (tmin > 0.f ? tmin : 0.f)
+                    : KokkosExt::ArithmeticTraits::infinity<float>::value;
```

Comment:
This is not what a distance between ray and box is. IMO, the distance between Ray and Box is always well defined and finite. Instead, this looks like "Distance from the ray origin to the box if they intersect, and infinity otherwise". So I would call it something else.

Reply:
I don't see the problem with the definition proposed here.  `inf` denotes "no distance".  What am I missing?

Reply:
I think @aprokop doesn't like that the definition here doesn't match the default "closest distance between two objects" which is not what I need here.

Reply:
Yes. `distance` for me is the distance between two geometric objects. As such, distance from a valid ray to a valid box is always finite.

Reply:
Any suggestions on how I can specify what I need/want using our current interface then?

Reply:
It's hard for me to say what you and @dalg24  are doing without seeing the code. But you could still use the function, just not name is `distance`.

Reply:
For ray-tracing, I need to know what the first object is a ray intersects and a `nearest` query is doing exactly that. More precisely, I don't want the distance between a Ray and an object, but the distance between a point and an object with respect to a given direction.
The `nearest` traversal uses `distance` to order matches and thus I need to define `distance` in order to use it.

Reply:
So I understand how that would allow you to find the closest object that ray intersects with. But I don't see how that would allow you to order all objects that a ray intersects with in increasing order distance. I thought for that you would need additional structures, as you would essentially need to maintain possibly unbounded number of intersections.

I also think we need to start using the proper terminology to describe different algorithms. I think following [this](https://onlinelibrary.wiley.com/doi/abs/10.1111/cgf.142662) paper could be a good choice. In Section 6.1, they establish several terms:
- First hit traversal (closest hit test)
  Finding the nearest object in the direction of a ray from its origin
- Any-hit traversal (occlusion test)
  Finding any object in a specific direction from a ray origin (and within a specified range)
- Multi-hit traversal
  Finding one or more primitives that intersect with a ray (k closest intersections from a viewpoint or all intersections)

So, in these terms, is the problem you are considering here first hit traversal, or multi-hit traversal?

Reply:
At this point, I want "First hit traversal". I don't think we need anything else for a reasonable proxy.

Reply:
OK, that makes more sense. I thought you were considering the problem that the university team brought up a few months ago, where you wanted to do all hits but ordered.